Update `actual_file_to_check` with rendered `path`

remem9527 opened the pull request:
The new added attribute `self.actual_file_to_check` is not one of template_fields and is initiated in `self.__init__()`. So, it need to be updated after `self.path` is rendered. Otherwise, the template in `self.path` will not work.

@alexkruc

remem9527 committed:
Update `actual_file_to_check` with rendered `path`

The new added attribute `self.actual_file_to_check` is not one of template_fields and is initiated in `self.__init__()`. So, it need to be updated after `self.path` is rendered. Otherwise, the template in `self.path` will not work.

josh-fell commented on airflow/providers/sftp/sensors/sftp.py:
```diff
@@ -71,6 +71,8 @@ def poke(self, context: 'Context') -> bool:
             else:
                 return False
+        else:
+            self.actual_file_to_check = self.path
```
Good catch! IMO it's probably not worth having `actual_file_to_check` be an instance attr since it's not used outside of the `poke()` method. WDYT? If you agree, it should be deleted from `__init__()` as well.

remem9527 replied:
Maybe we want to access one of the filenames that matches `self.file_pattern`, when `self.file_pattern` is not empty? I am not sure about that. We probably need some input from the committer of file_pattern supporting.
My true opinion is that the implementation about file_pattern is not very elegant. The variable `self.path` means a **file** path when `self.file_pattern` is empty. But when `self.file_pattern` is specified, the variable `self.path` is treated as a **folder** path.

alexkruc replied:
Indeed a good catch! :) 
I agree that `actual_file_to_check` can be deleted from the `__init__()` method, it's not really needed there, and it was my mistake to leave it there 😅 
Perhaps the best approach is to initiate the variable inside the `poke()` method itself.
Or, because it is a user-facing attribute, maybe it's ok to add this to the `template_fields`? WDYT? 

About the API itself, when I've added this feature, I saw `self.path` variable to be a strict and complete file location, but when you ask for a pattern, it's "easier" to understand that you look for a pattern, in a specific path.
So for example, if you want to check if csv reports are done and placed in the correct folder, you will give the folder with the `*.csv` pattern for poking.

But if you think that this API is not good, I guess it's ok also to change it :)

remem9527 replied:
Personally, I think it would be very cool if I could use sensors like this:
```
sensor_task = SFTPSensor(task_id='task_id', sftp_conn_id='sftp_conn_id', path='data/folder/datafile_{{ds}}_part_?.[ct]sv')
```
Also, this supporting should be implemented on some base class, so that all the file sensors could be supported.

josh-fell replied:
❤️ the discussion here! I think they are all great suggestions, but they don't necessarily need to be part of this PR. Definitely feel free to fire over PRs and we can discuss more.

>Perhaps the best approach is to initiate the variable inside the poke() method itself.

+1

remem9527 replied:
Submitted another commit accordingly. Closing this conversation

josh-fell commented on airflow/providers/sftp/sensors/sftp.py:
```diff
@@ -71,6 +71,8 @@ def poke(self, context: 'Context') -> bool:
             else:
                 return False
+        else:
+            self.actual_file_to_check = self.path
```
Good catch! IMO it's probably not worth having `actual_file_to_check` be an instance attr since it's not used outside of the `poke()` method. WDYT? If you agree, it should be deleted from `__init__()` as well.

remem9527 replied:
Duplicated. Closing it.

remem9527 committed:
Initiate `actual_file_to_check` in `poke()`

It is not necessary for `self.actual_file_to_check` to be an attribute. Deleted it from `__init__`. Replaced it with a local variable in `SFTPSensor.poker()`.
